### PR TITLE
added script to unbind and delete all services

### PR DESCRIPTION
**Pull request by jackjoy-gds**
What
----

This script can be used in isolation, or as part of the destroy-cloudfoundry pipeline to iterate over the orgs/spaces/apps/bindings in any targeted cloud foundry platform and will unbind any bound services before deleting them.
The PR also includes the necessary changes to the destroy-cf pipeline yaml

How to review
-------------

Run the script locally against a dev env using the test and or dryrun mode
Configure a dev pipeline to use this branch and trigger a destroy job in a dev env 

---

🚨⚠️ Please do not merge this pull request via the GitHub UI ⚠️🚨

### CODE REVIEW

File: scripts/unbind-and-delete-all-services.sh [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+
+delete_service() {
+  local org_name
+  local space_name
+  local service_instance_name
+  org_name="${1}"
+  space_name="${2}"
+  service_instance_name="${3}"
+  get_cf_target "${org_name}" "${space_name}"
+  if ! cf delete-service -w -f "${service_instance_name}" >/dev/null 2>&1; then
```
Comment (corlettb) [inline scripts/unbind-and-delete-all-services.sh:303]: This wait option will push pipeline time out. Consider running these in parallel.

File: scripts/unbind-and-delete-all-services.sh [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+set -e -u -o pipefail
+shopt -s inherit_errexit
+IFS=$'\n'
+
+#### config for test modes
+test_org_count="1"
+test_space_count="1"
+test_deploy_count="1"
+test_prefix="destroy-tests"
+test_dir="/tmp/${test_prefix}"
```
Comment (corlettb) [inline scripts/unbind-and-delete-all-services.sh:21]: Any reason you don't want to use the traditional
```
mktemp -d -t <prefix>
```
?

File: scripts/unbind-and-delete-all-services.sh [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+
+cf_curl() {
+  local url
+  url="${1}"
+  local json
+  json=""
+  if ! cf target >/dev/null 2>&1; then
+    print_error "not logged in"
+  elif ! json=$(cf curl "${url}" 2>&1); then
+    print_error "cf curl failed"
```
Comment (corlettb) [inline scripts/unbind-and-delete-all-services.sh:152]: We are masking the real error here and in lots of other places. From a pipeline point of view I want to know why it failed.

File: concourse/pipelines/destroy-cloudfoundry.yml [hunk 4]
```diff
@@ -136,41 +136,39 @@ jobs:
...
 
-      - task: remove-auditor-db
-        tags: [colocated-with-web]
-        file: paas-cf/concourse/tasks/remove-db.yml
-        params:
-          ORG: admin
-          SPACE: billing
-          DB_NAME: auditor-db
-          BOUND_APPS: paas-auditor
+                  ./paas-cf/scripts/unbind-and-delete-all-services.sh execute
```
Comment (corlettb) [inline concourse/pipelines/destroy-cloudfoundry.yml:171]: Script and pipeline contains no timeouts or retries. Do we want to consider this?

File: scripts/unbind-and-delete-all-services.sh [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+
+get_test_orgs() {
+  local org_num
+  local org_name
+  org_name=""
+  for org_num in $(seq 1 "${test_org_count}"); do
+    org_name="${org_name}${test_prefix}-org-${org_num},"
+  done
+  sed 's/,$//' < <(echo "${org_name}")
+}
```
Comment (corlettb) [inline scripts/unbind-and-delete-all-services.sh:133]: I don't think I'm a fan of mixing all the test code and functionality. You should consider a testing framework like bats. Here is a quick start guide, https://www.baeldung.com/linux/testing-bash-scripts-bats.

File: scripts/unbind-and-delete-all-services.sh (hunk 1 above)
Comment (jackjoy-gds) [reply scripts/unbind-and-delete-all-services.sh]: will consider in next iteration

File: scripts/unbind-and-delete-all-services.sh (hunk 5 above)
Comment (jackjoy-gds) [reply scripts/unbind-and-delete-all-services.sh]: will consider in future

**Review by corlettb — APPROVED** [review]

**Pull request merged**